### PR TITLE
New MXE packages with focus on security

### PR DESCRIPTION
**Pull request by hfmanson**
We ported these libraries for the sake of our [TLS Pool](http://tlspool.arpa2.net/) project. The daemon and libtlspool are also included and make it incredibly simple to turn a connection into a TLS-protected connection.

Support packages ported for this purpose are:
- p11-kit, a tool to use standardised pkcs11: URIs to access cryptographic tokens;
- unbound / ldns, resolver libraries that implement DNSSEC
- GnuTLS has been upgraded to use p11-kit
- Quick DER and asn2quickder are straightforward handling libraries for DER representation of ASN.1

We also introduced `--content-dispositiion` for `wget`, to make it possible to use GitHub releases.

### CODE REVIEW

File: docs/build-matrix.html [hunk 1]
```diff
@@ -3478,6 +3478,16 @@
...
+ <th class="row">quick-der</th>
+ <td>0.1-RC2</td>
+      <td class="supported">&#x2713;</td>
+      <td class="supported">&#x2713;</td>
+      <td class="supported">&#x2713;</td>
+      <td class="supported">&#x2713;</td>
+  <td class="unsupported">&#x2717;</td>
+ </tr>
+
+<tr>
```
Comment (starius) [inline docs/build-matrix.html]: Humans are not expected to update this file. It is updated by @MXEBot

File: src/asn2quickder.mk [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := 0.6-RC2
+$(PKG)_CHECKSUM := 0d92fdec9ed6f7f12e1994196e46c73e0721ff560eebde59ca8f7ef21fd03068
+$(PKG)_SUBDIR   := $(PKG)-version-$($(PKG)_VERSION)
+$(PKG)_FILE     := version-$($(PKG)_VERSION).tar.gz
+$(PKG)_URL	:= https://github.com/vanrein/asn2quickder/archive/$($(PKG)_FILE)
+$(PKG)_DEPS     := gcc
+
+define $(PKG)_UPDATE
+    $(WGET) -q -O- 'https://github.com/vanrein/asn2quickder/tags' | \
```
Comment (starius) [inline src/asn2quickder.mk]: Use `MXE_GET_GITHUB_TAGS`.

File: src/asn2quickder.mk [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# This file is part of MXE.
+# See index.html for further information.
+
+PKG             := asn2quickder
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := 0.6-RC2
+$(PKG)_CHECKSUM := 0d92fdec9ed6f7f12e1994196e46c73e0721ff560eebde59ca8f7ef21fd03068
+$(PKG)_SUBDIR   := $(PKG)-version-$($(PKG)_VERSION)
+$(PKG)_FILE     := version-$($(PKG)_VERSION).tar.gz
+$(PKG)_URL	:= https://github.com/vanrein/asn2quickder/archive/$($(PKG)_FILE)
```
Comment (starius) [inline src/asn2quickder.mk]: Fix alignment of `:=`.

File: src/asn2quickder.mk [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# This file is part of MXE.
+# See index.html for further information.
```
Comment (starius) [inline src/asn2quickder.mk]: Replace with the following:
```
This file is part of MXE. See LICENSE.md for licensing information.
```

File: src/asn2quickder.mk [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+$(PKG)_MAKE_OPTS = \
+	PREFIX='$(PREFIX)/$(BUILD)' \
+	CROSS_SUFFIX='$(TARGET)-' \
+	CC='$(TARGET)-gcc' \
+	AR='$(TARGET)-ar' \
+	HOSTCC='$(BUILD_CC)'
+
+define $(PKG)_BUILD
+    $(MAKE) PREFIX='$(PREFIX)/$(TARGET)' -C '$(1)' -j '$(JOBS)' all install $($(PKG)_MAKE_OPTS)
```
Comment (starius) [inline src/asn2quickder.mk]: Please replace `$(1)` with `$(SOURCE_DIR)`

File: src/gnutls-2-fixes2.patch [hunk 6]
```diff
@@ -0,0 +1,20 @@
+--- a/libdane/dane-params.c	2016-07-12 18:03:22.691584000 +0200
```
Comment (starius) [inline src/gnutls-2-fixes2.patch]: Please format patches using patch-tool-mxe. See https://github.com/mxe/mxe/pull/1277#issuecomment-257083463 for explanation. Provide commit messages with reasoning of patches.

File: src/gnutls.mk [hunk 7]
```diff
@@ -28,8 +28,8 @@ define $(PKG)_BUILD
         --enable-local-libopts \
         --with-included-libtasn1 \
         --with-libregex-libs="-lgnurx" \
-        --without-p11-kit \
         --disable-silent-rules \
+		--with-unbound-root-key-file="C:\\\\Program Files\\\\Unbound\\\\root.key"
```
Comment (starius) [inline src/gnutls.mk]: 1. Please fix alignment.
2. About "C:\\\\Program Files\\\\Unbound\\\\root.key" path... Can you provide more information on this path? Can we avoid hardcoding it in the build rules?

File: src/asn2quickder.mk [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+define $(PKG)_UPDATE
+    $(WGET) -q -O- 'https://github.com/vanrein/asn2quickder/tags' | \
+    grep '<a href="/vanrein/asn2quickder/archive/' | \
+    $(SED) -n 's,.*href="/vanrein/asn2quickder/archive/version-\([0-9][^"]*\)\.tar.*,\1,p' | \
+    head -1
+endef
+
+$(PKG)_MAKE_OPTS = \
+	PREFIX='$(PREFIX)/$(BUILD)' \
```
Comment (starius) [inline src/asn2quickder.mk]: Please use spaces instead of tabs here.

File: Makefile [hunk 9]
```diff
@@ -238,7 +238,7 @@ DOWNLOAD_PKG_ARCHIVE = \
         true\
     $(else),\
         mkdir -p '$(PKG_DIR)' && ( \
-            $(WGET) -T 30 -t 3 -O- '$($(1)_URL)' \
+            $(WGET) --content-disposition -T 30 -t 3 -O- '$($(1)_URL)' \
```
Comment (starius) [inline Makefile]: Why is it needed?

From `man wget`:

> This option is useful for some file-downloading CGI programs that use "Content-Disposition" headers to describe what the name of a downloaded file should be.

Our downloading code does not use filename provided by a server.

File: src/libtasn1.mk [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+define $(PKG)_UPDATE
+    echo 'TODO: write update script for $(PKG).' >&2;
+    echo $($(PKG)_VERSION)
+endef
+
+define $(PKG)_BUILD
+    cd '$(1)' && ./configure \
+        $(MXE_CONFIGURE_OPTS) \
+#        --disable-rpath \
```
Comment (starius) [inline src/libtasn1.mk]: This line and 3 following lines seem to be commented out. Can you remove or uncomment them?

File: src/p11-kit.mk [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+define $(PKG)_UPDATE
+    echo 'TODO: write update script for $(PKG).' >&2;
+    echo $($(PKG)_VERSION)
+endef
+
+define $(PKG)_BUILD
+    cd '$(1)' && ./configure \
+        $(MXE_CONFIGURE_OPTS) \
+#        --disable-rpath \
```
Comment (starius) [inline src/p11-kit.mk]: This line and 3 following lines seem to be commented out. Can you remove or uncomment them?

File: src/quick-der.mk [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+endef
+
+$(PKG)_MAKE_OPTS = \
+	PREFIX='$(PREFIX)/$(TARGET)' \
+	CROSS_SUFFIX='$(TARGET)-' \
+	CC='$(TARGET)-gcc' \
+	AR='$(TARGET)-ar' \
+	ASN2QUICKDER_CMD='$(PREFIX)/$(BUILD)/bin/asn2quickder' \
+	HOSTCC='$(BUILD_CC)' \
+    WINVER=0x0600 \
```
Comment (starius) [inline src/quick-der.mk]: Alignment

File: src/tlspool-gui.mk [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+$(PKG)_URL      := https://github.com/amarsman/$(PKG)/archive/v$($(PKG)_VERSION).tar.gz
+$(PKG)_DEPS     := qtsvg tlspool
+
+define $(PKG)_UPDATE
+    echo 'TODO: write update script for $(PKG).' >&2;
+    echo $($(PKG)_VERSION)
+endef
+
+define $(PKG)_BUILD
+    mkdir '$(1)/build'
```
Comment (starius) [inline src/tlspool-gui.mk]: Use `$(BUILD_DIR)` instead.

File: src/tlspool.mk [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+$(PKG)_MAKE_OPTS = \
+        PREFIX='$(PREFIX)/$(TARGET)' \
+        CROSS_SUFFIX='$(TARGET)-' \
+        CC='$(TARGET)-gcc' \
+		PKG_CONFIG='$(TARGET)-pkg-config' \
+		DLLTOOL='$(TARGET)-dlltool' \
+        HOSTCC='$(BUILD_CC)' \
+        WINVER=0x0600 \
+		SBIN=bin
```
Comment (starius) [inline src/tlspool.mk]: Alignment

File: src/tlspool.mk [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+define $(PKG)_BUILD
+    $(MAKE)  -C '$(1)/src' -j '$(JOBS)' install $($(PKG)_MAKE_OPTS)
```
Comment (starius) [inline src/tlspool.mk]: Extra space after `$(MAKE)`

File: src/unbound.mk [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+# See index.html for further information.
+
+PKG             := unbound
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := 1.5.10_20160712
+$(PKG)_CHECKSUM := 50b8ea88b88d6db899fcb43cd35497791b729fee5d6690b6205f89497c0575e1
+$(PKG)_SUBDIR   := unbound-$($(PKG)_VERSION)
+$(PKG)_FILE     := unbound-$($(PKG)_VERSION).tar.gz
+$(PKG)_URL      := http://www.nlnetlabs.nl/~wouter/$($(PKG)_FILE)
+# $(PKG)_URL_2    := .../$($(PKG)_FILE)
```
Comment (starius) [inline src/unbound.mk]: Remove or uncomment.

File: src/unbound.mk [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+$(PKG)_DEPS     := gcc openssl zlib expat
+
+
```
Comment (starius) [inline src/unbound.mk]: Remove extra empty lines.

File: src/unbound.mk [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+    $(WGET) -q -O- 'https://unbound.net/downloads/' | \
+    grep 'unbound-[0-9]' | \
+    $(SED) -n 's,.*unbound-\([0-9][^>]*\)\.tar\.gz.*,\1,p' | \
+    grep -v 'rc[0-9]' | \
+    $(SORT) -Vr | \
+    head -1
+endef
+
+define $(PKG)_BUILD
+    cd '$(1)' && ./configure \
```
Comment (starius) [inline src/unbound.mk]:
```
cd '$(BUILD_DIR)' && $(SOURCE_DIR)/configure
    $(MXE_CONFIGURE_OPTS)
```

File: src/unbound.mk [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+		$(MXE_CONFIGURE_OPTS) \
+		--disable-flto \
+		--enable-shared \
```
Comment (starius) [inline src/unbound.mk]: Shared/static should depend on `$(TARGET)`. Macro `$(MXE_CONFIGURE_OPTS)` automates this.

File: src/unbound.mk [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+		--without-pthreads \
+		--with-libexpat=$(PREFIX)/$(TARGET) \
+		--with-conf_file="'C:\Program Files (x86)\Unbound\service.conf'" \
```
Comment (starius) [inline src/unbound.mk]: Can you provide more information on this path? Can we avoid hardcoding it in the build rules?

**Comment by starius** [comment]
On addition of wv, enchant and abiword:
1. This seems to be unrelated to previous commits of this pull request. Please move to a separate pull request.
2. Core MXE packages (`src/*.mk`) are libraries, not applications. While wv and enchant are libraries and they are suitable for MXE core, the package `abiword` is an application and should be moved to plugins/apps plugin. See https://github.com/mxe/mxe/tree/master/plugins for more information about plugins.

**Comment by starius** [comment]
I see some commits from master branch in this pull request.
Can you create new pull request with new commits only, please?

**Comment by tonytheodore** [comment]
Based on:
```
git checkout cc0440369837
git checkout -b arpa2-master
git pull https://github.com/arpa2/mxe.git master
ls -1 src/*.mk > arpa2
git checkout master
ls -1 src/*.mk > master
grep -v -f master arpa2
```
I think these are the new packages:
```
src/abiword.mk
src/asn2quickder.mk
src/enchant.mk
src/ldns.mk
src/libtasn1.mk
src/p11-kit.mk
src/quick-der.mk
src/tlspool-gui.mk
src/tlspool.mk
src/unbound.mk
src/wv.mk
```

**Comment by tonytheodore** [comment]
Replaced with #1960

**Pull request closed without merging**